File: 37.解数独.js
/*
 * @lc app=leetcode.cn id=37 lang=javascript
 *
 * [37] 解数独
 */

// @lc code=start
/**
 * @param {character[][]} board
 * @return {void} Do not return anything, modify board in-place instead.
 */
 var solveSudoku = (board) => {
    X.fill(0);
    Y.fill(0);
    Z.fill(0);
    initBoard(board);
    solve(board, 0, 0);
}

const X = Array(9).fill(0);
const Y = Array(9).fill(0);
const Z = Array(9).fill(0);

function initBoard(board) {
    for (let i = 0; i < board.length; i++) {
        for (let j = 0; j < board[0].length; j++) {
            if (board[i][j] === ".") {
                continue;
            } else {
                setCheck(board[i][j] - "0", i, j);
            }
        }
    }
}

function setCheck(num, x, y) {
    X[x] ^= (1 << num);
    Y[y] ^= (1 << num);
    Z[Math.floor(x / 3) * 3 + Math.floor(y / 3)] ^= (1 << num); 
}

function check(num, x, y) {
    if (X[x] & (1 << num)) {
        return false;
    }
    if (Y[y] & (1 << num)) {
        return false;
    }
    if (Z[Math.floor(x / 3) * 3 + Math.floor(y / 3)] & (1 << num)) {
        return false;
    }
    return true;
}

function nextX(x, y) {
    if (y === 8) {
        x++;
    }
    return x;
}

function nextY(y) {
    if (y === 8) {
        y = 0;
    } else {
        y++;
    }
    return y;
}

function solve(board, x, y) {
    if (x === 9) {
        return true;
    }
    if (board[x][y] !== '.') return solve(board, nextX(x,y), nextY(y));
    for (let i = 1; i <= 9; i++) {
        if (!check(i, x, y)) {
            continue;
        }
        board[x][y] = i + "";
        setCheck(i, x, y);
        if (solve(board, nextX(x, y), nextY(y))) {
            return true;
        }
        board[x][y] = ".";
        setCheck(i, x, y);
    }
    return false;
}
// @lc code=end

File: 37.解数独.rs
/*
 * @lc app=leetcode.cn id=37 lang=rust
 *
 * [37] 解数独
 */

// @lc code=start
impl Solution {
    pub fn solve_sudoku(board: &mut Vec<Vec<char>>) {
        let mut c = [[false; 9]; 9];
        let mut h = [[false; 9]; 9];
        let mut v = [[false; 9]; 9];
        for y in 0..9 {
            for x in 0..9 {
                if let Some(n) = board[y][x].to_digit(10) {
                    let p = (n - 1) as usize;
                    let sub_board = y / 3 * 3 + x / 3;
                    c[sub_board][p] = true;
                    h[y][p] = true;
                    v[x][p] = true;
                }
            }
        }
        Solution::fill(board, 0, 0, &mut c, &mut h, &mut v);
    }

    pub fn fill(
        board: &mut Vec<Vec<char>>,
        mut x: usize,
        mut y: usize,
        c: &mut [[bool; 9]; 9],
        h: &mut [[bool; 9]; 9],
        v: &mut [[bool; 9]; 9],
    ) -> bool {
        while board[y][x] != '.' {
            if x == 8 {
                if y == 8 {
                    return true;
                }
                y += 1;
                x = 0;
            } else {
                x += 1;
            }
        }
        if board[y][x] == '.' {
            let sub_board = y / 3 * 3 + x / 3;
            for num in 0..9 {
                if !c[sub_board][num] && !h[y][num] && !v[x][num] {
                    board[y][x] = char::from_digit(num as u32 + 1, 10).unwrap();
                    c[sub_board][num] = true;
                    h[y][num] = true;
                    v[x][num] = true;
                    if Solution::fill(board, x, y, c, h, v) {
                        return true;
                    }
                    board[y][x] = '.';
                    c[sub_board][num] = false;
                    h[y][num] = false;
                    v[x][num] = false;
                }
            }
        }

        false
    }
}
// @lc code=end

